src/AdoPat/PatClient.cs

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+namespace Microsoft.Authentication.AdoPat
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.VisualStudio.Services.DelegatedAuthorization;
+
+    /// <summary>
+    /// An abstraction over common operations with the Azure DevOps PAT Lifecycle Management REST API.
+    /// </summary>
+    public class PatClient
+    {
+        private const int PageSize = 100; // Using the maximum allowable page size allows us to reduce HTTP calls.
+        private const bool AllOrgs = false; // Azure DevOps have recommended we always set this to false.
+
+        private ITokensHttpClientWrapper client;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="PatClient"/> class.
+        /// </summary>
+        /// <param name="client">Any class which implements the <see cref="ITokensHttpClientWrapper"/> interface.</param>
+        public PatClient(ITokensHttpClientWrapper client)
+        {
+            this.client = client;
+        }
+
+        /// <summary>
+        /// Creates a new PAT.
+        /// </summary>
+        /// <param name="patTokenCreateRequest">
+        /// XXX: Not documented by Visual Studio client libraries. Should we depend on this or
+        /// should this method use the corresponding primitives as arguments? Not yet clear.

```

Comment:
Given that this is not documented by client libraries, how about we create a separate private method like `BuildPatTokenCreateRequest()` and use that method at all the places we need this object. And then the caller doesn't really have to worry about what's a `patTokenCreateRequest` object and can pass primitives as arguments.

Reply:
A `PatTokenCreateRequest` is pretty simple to construct. We could add a private method, but it would be just as easy to create it in this method, I think. It has fields
```
string DisplayName
string Scope
DateTime? ValidTo
bool AllOrgs = false
```
The reason I didn't initially use the primitives is because this seems like a [data clump](https://refactoring.guru/smells/data-clumps) to me where all these parameters are bound up together and highly related.

I considered not leaking implementation details from a library we rely on, but I'm also returning `PatToken` elsewhere, which is another Visual Studio type. In that case and the case of `PatTokenCreateRequest`, however, those are both just data classes with no behavior on them, esentially just containers for data which belong together.

Since we're not publishing this as a C# library I've played a bit fast and loose by re-using external types instead of wrapping them in our own. Basically, what I'm asking here is whether that seems like an acceptable tradeoff.

Reply:
Got it. That makes things much clear. It seems reasonable tradeoff to me :)